AWS: Resync EBS code with GCE PD, misc fixes

justinsb opened the pull request:
The main change here is to re-sync the AWS EBS code to the GCE PD code.  Sadly they are a copy-and-paste of each other (I would like to start moving common functionality into a shared directory, but first things first).

The GCE PD code accumulated nice retry logic around volume mounting which we want in EBS, as it should help with some AWS EBS edge cases (primarily in pkg/volume/aws_ebs/aws_util.go)

There are also some fixes to make the mount-device clearer (which fixes a mistake/confusion I made before), and to handle deleting a volume which doesn't exist (which allows the tests to pass, because the tests delete volumes twice assuming idempotency)

justinsb committed:
AWS: Handle deleting volume that no longer exists

The tests in particular double-delete volumes, so we need to handle this
graciously.

saad-ali commented:
Thanks a lot for doing this @justinsb 

I'm actually in the process of modifying a lot of the GCE code to fix #15382. I will send out that PR in the morning (running soak tests overnight for verification). That PR will get rid of `operationmanager` and make attach/detach atomic operations (the operation manager tries to defer to pending detach, the new PR will have a mutex per PD and treat attach/detach as atomic operations). That should make attach/detach even more stable. But, it also means an ugly rebase for this PR--sorry about that!

Looking past 1.1, this design where the decision to attach and detach is decoupled in two separate asynchronous loops is inherently broken. I will push a design for one controller that decides whether a PD should be attached or detached; see #15524. That should fix all the nasty race conditions that result from the existing design as well as the "disk attached to downed nodes" issue (#14642).

justinsb commented:
@saad-ali no problem & thanks for the heads-up.  I guess I will hold off until your PR merges, and then I'll probably just redo this one.  I'm debating whether to split this PR up, but I worry that if we do that it'll make it harder to cherry-pick into 1.1 (if we choose to do that)

saad-ali commented:
Sent out PR #16053. I say a single PR is fine.

saad-ali commented:
FYI #16053 is in.

justinsb committed:
AWS: Update copy-paste of GCE PD code to latest version

We are (sadly) using a copy-and-paste of the GCE PD code for AWS EBS.
This code hasn't been updated in a while, and it seems that the GCE code
has some code to make volume mounting more robust that we should copy.

justinsb committed:
Fix typo in comment: succefully -> successfully

justinsb commented:
Reworked this with the latest GCE PD code.  After we get this merged, it would be great to refactor some of this copy-pasta into a common module, before it diverges again :-)

justinsb commented:
Rebased

justinsb commented:
@saad-ali I'd love to get this into 1.2.

The big advantages as I see them are that we get the fixes into AWS that GCE has made (good for AWS), and that by having GCE & AWS on similar code if we find a problem on the AWS side we can hopefully fix it for GCE also (good for GCE!).

What do you think?

saad-ali commented on pkg/volume/aws_ebs/aws_util.go:
```diff
@@ -17,47 +17,58 @@ limitations under the License.
+
+	glog.V(5).Infof("AttachAndMountDisk(...) called for PD %q. Awake and ready to execute. (globalPDPath=%q)\r\n", b.volumeID, globalPDPath)
+
+	xvdBefore, err := filepath.Glob(diskXVDPattern)
```
All the `udevadm` stuff was a workaround for a GCE PD issue. It shouldn't hurt, but since it is unnecessary, we should remove it.

saad-ali commented:
> The big advantages as I see them are that we get the fixes into AWS that GCE has made (good for AWS), and that by having GCE & AWS on similar code if we find a problem on the AWS side we can hopefully fix it for GCE also (good for GCE!).

> What do you think?

Totally agreed! Thanks for doing this. LGTM mod nit (udevadm stuff can be removed from the AWS side, or left in if we want the future factoring of GCE/AWS to be easier, up to you).

justinsb committed:
AWS EBS: HandleCrash moved to util/runtime

justinsb committed:
AWS: Switch arguments to AttachDisk/DetachDisk to match GCE

justinsb committed:
AWS EBS: Remove copied-and-pasted udevadm code

This is in GCE to deal with a GCE specific issue; there's no reason to
believe we need the same logic on AWS.

justinsb commented:
Weird e2e failure... investigating...  (not related to this PR I don't think)

```
Feb  3 13:24:51.648: INFO: Failed to delete cloud resources for service: 104.197.185.91 19482 (googleapi: Error 400: Invalid value for field 'region': 'us-central1-f'.  Unknown region., invalid)
STEP: stopping RC webserver in namespace e2e-tests-services-gb3yh
```

justinsb commented:
@k8s-bot test this

Tracking flake in #20599

justinsb commented:
Looks like k8s-merge-robot deleted my comment, but still ran the tests.  They passed this time!  The flake is #20599

justinsb commented:
Removed the udevadm code, and switched the arguments in AttachDisk/DetachDisk to match the same order as GCE to stop people getting confused like I did!

I've also run AWS e2e with this code, and it passes.

@saad-ali what do you think?  I'd love for this to get in to 1.2.

justinsb commented:
@k8s-bot test this github flake: #20599

Although the e2e test succeeded, github isn't reflecting that above.

justinsb commented:
@k8s-bot unit test this github flake: #20599

Still missing the unit test

saad-ali commented:
Thanks @justinsb. Will take a look today

saad-ali commented:
LGTM
